File: services/grib.go
package services

import (
	"fmt"
	"github.com/xairline/xa-snow/utils/logger"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"
)

// grib service
type GribService interface {
	IsReady() bool                                                                              // ready to retrieve values
	DownloadAndProcessGribFile(sys_time bool, day, month, hour int) (error, DepthMap, DepthMap) // -> err, gribSnow, coastalSnow
	GetSnowDepth(lat, lon float32) float32
	convertGribToCsv(snow_csv_name string)
	downloadGribFile(sys_time bool, day, month, hour int) (string, error)
	getDownloadUrl(sys_time bool, timeUTC time.Time) (string, time.Time, int)
	SetNotReady()
}

type gribService struct {
	ready          bool
	Logger         logger.Logger
	gribFilePath   string
	gribFileFolder string
	binPath        string
	cs             CoastService
	SnowDm         DepthMap
}

func (g *gribService) SetNotReady() {
	g.ready = false
}

var gribSvcLock = &sync.Mutex{}
var gribSvc GribService

func NewGribService(logger logger.Logger, dir string, binPath string, cs CoastService) GribService {
	if gribSvc != nil {
		logger.Info("Grib SVC has been initialized already")
		return gribSvc
	} else {
		logger.Info("Grib SVC: initializing")
		gribSvcLock.Lock()
		defer gribSvcLock.Unlock()
		logger.Infof("Grib SVC: initializing with folder %s", dir)
		gribSvc = &gribService{
			Logger:         logger,
			gribFileFolder: dir,
			gribFilePath:   "",
			binPath:        binPath,
			cs:             cs,
		}
		// make sure grib file folder exists
		if _, err := os.Stat(dir); os.IsNotExist(err) {
			os.MkdirAll(dir, os.ModePerm)
		}
		return gribSvc
	}
}

func (g *gribService) IsReady() bool {
	return g.ready
}

func (g *gribService) GetSnowDepth(lat, lon float32) float32 {
	if !g.ready {
		g.Logger.Error("Get called and service is not ready!")
		return 0.0
	}

	return g.SnowDm.Get(lon, lat)
}

func (g *gribService) DownloadAndProcessGribFile(sys_time bool, month, day, hour int) (error, DepthMap, DepthMap) {
	file_override := 0

	snow_csv_file := "snod.csv"

	tmp := os.Getenv("USE_SNOD_CSV")
	if tmp != "" {
		snow_csv_file = tmp
		file_override++
	}

	var gribFilename string
	var err error

	if file_override < 1 {
		// download grib file
		gribFilename, err = g.downloadGribFile(sys_time, day, month, hour)
		if err != nil {
			return err, nil, nil
		}
		// convert grib file to csv files
		g.convertGribToCsv("snod.csv")
	}

	gribSnow := NewDepthMap()
	gribSnow.LoadCsv(snow_csv_file)

	// remove old grib files
	err = g.removeOldGribFiles(gribFilename)
	if err != nil {
		return err, nil, nil
	}

	coastalSnow := ElsaOnTheCoast(gribSnow)
	g.SnowDm = coastalSnow
	g.ready = true
	return nil, gribSnow, coastalSnow
}

func (g *gribService) getDownloadUrl(sys_time bool, timeUTC time.Time) (string, time.Time, int) {
	g.Logger.Infof("timeUTC:  %s", timeUTC.String())
	ctimeUTC := timeUTC.Add(-4*time.Hour - 25*time.Minute) // Adjusted time considering publish delay
	g.Logger.Infof("ctimeUTC: %s", ctimeUTC.String())
	cycles := []int{0, 6, 12, 18}
	var cycle int
	for _, cycle_ := range cycles {
		if ctimeUTC.Hour() >= cycle_ {
			cycle = cycle_
		}
	}

	adjs := 0
	if ctimeUTC.Day() != timeUTC.Day() {
		adjs = 24
	}
	forecast := (adjs + timeUTC.Hour() - cycle) / 3 * 3

	cycleDate := fmt.Sprintf("%d%02d%02d", ctimeUTC.Year(), ctimeUTC.Month(), ctimeUTC.Day())

	if sys_time {
		filename := fmt.Sprintf("gfs.t%02dz.pgrb2.0p25.f0%02d", cycle, forecast)
		g.Logger.Infof("NOAA Filename: %s, %d, %d", filename, cycle, forecast)
		url := fmt.Sprintf("https://nomads.ncep.noaa.gov/cgi-bin/filter_gfs_0p25.pl?dir=%%2Fgfs.%s%%2F%02d%%2Fatmos&file=%s&var_SNOD=on&all_lev=on", cycleDate, cycle, filename)
		return url, ctimeUTC, cycle
	} else {
		forecast = 6 // TODO: for now
		filename := fmt.Sprintf("gfs.0p25.%s%02d.f0%02d.grib2", cycleDate, cycle, forecast)
		g.Logger.Infof("GITHUB Filename: %s, %d, %d", filename, cycle, forecast)
		url := fmt.Sprintf("https://github.com/xairline/weather-data/releases/download/daily/%s", filename)
		return url, ctimeUTC, cycle
	}

	//return fmt.Sprintf("https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.%s/%02d/atmos/%s", cycleDate, cycle, filename)
}

func (g *gribService) convertGribToCsv(snow_csv_name string) {
	g.Logger.Infof("Pre-processing GRIB file to CSV: '%s'", snow_csv_name)
	//get current OS
	myOs := runtime.GOOS
	var executablePath string
	if myOs == "windows" {
		executablePath = filepath.Join(g.binPath, "WIN32wgrib2.exe")
	}
	if myOs == "linux" {
		executablePath = filepath.Join(g.binPath, "linux-wgrib2")
	}
	if myOs == "darwin" {
		executablePath = filepath.Join(g.binPath, "OSX11wgrib2")
	}
	// export grib file to csv
	// 0:3600:0.1 means scan longitude from 0, 3600 steps with step 0.1 degree
	// -90:1800:0.1 means scan latitude from -90, 1800 steps with step 0.1 degree
	cmd := exec.Command(executablePath, "-s", "-lola", "0:3600:0.1", "-90:1800:0.1", snow_csv_name, "spread", g.gribFilePath, "-match_fs", "SNOD")
	err := g.exec(cmd)
	if err != nil {
		g.Logger.Errorf("Error converting grib file: %v", err)
	}

	g.Logger.Info("Pre-processing GRIB file to CSV: Done")
}

// day, month, hour are in the local TZ
func (g *gribService) downloadGribFile(sys_time bool, day, month, hour int) (string, error) {
	g.Logger.Infof("downloadGribFile: Using system time: %t, month: %d, day: %d, hour: %d",
		sys_time, month, day, hour)

	now := time.Now()
	timeUTC := now.UTC()
	// Convert the provided day, month, and hour into a time.Time object
	providedTime := time.Date(now.Year(), time.Month(month), day, hour, 0, 0, 0, time.UTC)

	// Check if the provided time is within the last 24 hours
	if providedTime.After(now.Add(-24*time.Hour)) && providedTime.Before(now) && !sys_time {
		g.Logger.Infof("The provided time is within the last 24 hours. Using system time.")
		sys_time = true
	}

	if !sys_time {
		// historic mode

		loc := now.Location() // my TZ
		year := now.Year()

		m := int(now.Month())
		if (month > m) ||
			(month == m && day > now.Day()) ||
			(month == m && day == now.Day() && hour > now.Hour()) {
			// future month/day/hour -> use previous year
			year--
		}

		timeUTC = time.Date(year, time.Month(month), day, hour, 0, 0, 0, loc).UTC()
	}

	url, ctimeUTC, cycle := g.getDownloadUrl(sys_time, timeUTC)
	g.Logger.Infof("Downloading GRIB file from %s", url)
	// Get today's date in yyyy-mm-dd format
	today := ctimeUTC.Format("2006-01-02")
	// Create the filename with today's date
	filename := today + "_" + fmt.Sprintf("%d", cycle) + "_noaa.grib2"
	g.gribFilePath = filepath.Join(g.gribFileFolder, filename)
	g.Logger.Infof("GRIB file path: %s", g.gribFilePath)

	// if file does not exist, download
	if _, err := os.Stat(g.gribFilePath); err != nil {
		// Get the data
		resp, err := http.Get(url)
		if err != nil {
			g.Logger.Errorf("%v", err)
			return "", err
		}
		defer resp.Body.Close()

		// Create the file with the date in its name
		out, err := os.Create(g.gribFilePath)
		if err != nil {
			g.Logger.Errorf("%v", err)
			return "", err
		}
		defer out.Close()
		if resp.StatusCode != 200 {
			g.Logger.Errorf("Failed to download GRIB file: %s", resp.Status)
		}

		// Write the body to file
		_, err = io.Copy(out, resp.Body)
		if err != nil {
			g.Logger.Errorf("%v", err)
			return "", err
		}
		g.Logger.Info("GRIB File downloaded successfully")
	}
	return filename, nil
}

func (g *gribService) removeOldGribFiles(fileToKeep string) error {
	//Remove old .grib files
	g.Logger.Info("Removing old grib files")
	g.Logger.Infof("File to keep: %s", fileToKeep)
	g.Logger.Infof("Grib file folder: %s", g.gribFileFolder)
	err := filepath.Walk(g.gribFileFolder, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if strings.Contains(path, "Scenery") {
			return nil
		}

		// Check for files with .grib extension
		if strings.Contains(path, "_noaa.grib2") && !strings.Contains(path, fileToKeep) {
			err := os.Remove(path)
			if err != nil {
				g.Logger.Errorf("Error removing file:", path, err)
			} else {
				g.Logger.Infof("Removed: %s", path)
			}
		}

		return nil
	})
	if err != nil {
		g.Logger.Errorf("Error removing old grib files: %v", err)
		return err
	}
	return nil
}

File: grib.cpp
//
//    X Airline Snow: show accumulated snow in X-Plane's world
//
//    Copyright (C) 2025  Zodiac1214
//    Copyright (C) 2025  Holger Teutsch
//
//    This library is free software; you can redistribute it and/or
//    modify it under the terms of the GNU Lesser General Public
//    License as published by the Free Software Foundation; either
//    version 2.1 of the License, or (at your option) any later version.
//
//    This library is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
//    Lesser General Public License for more details.
//
//    You should have received a copy of the GNU Lesser General Public
//    License along with this library; if not, write to the Free Software
//    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
//    USA
//

#include <cstdio>
#include <ctime>
#include <array>
#include <string>
#include <mutex>
#include <future>
#include <chrono>
#include <filesystem>
#include <stdexcept>

#include "xa-snow.h"

// in:  user specified time
// out: url, cycle time, cycle num
static std::tuple<std::string, std::tm, int>
GetDownloadUrl(bool sys_time, const std::tm utime_utc)
{
    // Adjusted time considering publish delay
    std::tm ctime_utc = utime_utc;
    ctime_utc.tm_hour -= 4;
    ctime_utc.tm_min -= 25;
    std::mktime(&ctime_utc); // Normalize the time structure

    char buffer[1000];
    std::strftime(buffer, sizeof(buffer), "%Y-%m-%d %H:%M:%S", &ctime_utc);
    log_msg("adjusted utime_utc: %s", buffer);

    const static std::array<int, 4> cycles = {0, 6, 12, 18};
    int cycle = 0;
    for (int c : cycles) {
        if (ctime_utc.tm_hour >= c) {
            cycle = c;
        }
    }

    int adjs = 0;
    if (ctime_utc.tm_mday != utime_utc.tm_mday) {
        adjs = 24;
    }
    int forecast = (adjs + utime_utc.tm_hour - cycle) / 3 * 3;

    snprintf(buffer, sizeof(buffer), "%d%02d%02d", ctime_utc.tm_year + 1900, ctime_utc.tm_mon + 1, ctime_utc.tm_mday);
    std::string cycleDate(buffer);

    if (sys_time) {
        snprintf(buffer, sizeof(buffer), "gfs.t%02dz.pgrb2.0p25.f0%02d", cycle, forecast);
        std::string filename(buffer);
        log_msg("NOAA Filename: '%s', %d, %d", filename.c_str(), cycle, forecast);

        snprintf(buffer, sizeof(buffer), "https://nomads.ncep.noaa.gov/cgi-bin/filter_gfs_0p25.pl?dir=%%2Fgfs.%s%%2F%02d%%2Fatmos&file=%s&var_SNOD=on&all_lev=on", cycleDate.c_str(), cycle, filename.c_str());
        return {buffer, ctime_utc, cycle};
    } else {
        forecast = 6; // TODO: for now
        snprintf(buffer, sizeof(buffer), "gfs.0p25.%s%02d.f0%02d.grib2", cycleDate.c_str(), cycle, forecast);
        std::string filename(buffer);
        log_msg("GITHUB Filename: '%s', %d, %d", filename.c_str(), cycle, forecast);

        snprintf(buffer, sizeof(buffer), "https://github.com/xairline/weather-data/releases/download/daily/%s", filename.c_str());
        return {buffer, ctime_utc, cycle};
    }
}

static std::string
DownloadGribFile(bool sys_time, int day, int month, int hour)
{
    log_msg("downloadGribFile: Using system time: %d, month: %d, day: %d, hour: %d", sys_time, day, month, hour);

    std::time_t now = std::time(nullptr);
    std::tm now_tm = *std::localtime(&now);
    std::time_t provided_time = now;


    if (! sys_time) {
        std::tm provided_tm = now_tm;
        provided_tm.tm_mon = month - 1; // Adjust month
        provided_tm.tm_mday = day;
        provided_tm.tm_hour = hour;
        provided_tm.tm_min = 0;
        provided_tm.tm_sec = 0;

        provided_time = std::mktime(&provided_tm);
        std::time_t twenty_four_hours_ago = now - 24 * 60 * 60;

        // Check if the provided time is not within the last 24 hours
        if (provided_time <= twenty_four_hours_ago || provided_time > now) {
            // historic mode
            int year = now_tm.tm_year;
            int current_month = now_tm.tm_mon + 1;
            int current_day = now_tm.tm_mday;
            int current_hour = now_tm.tm_hour;

            if ((month > current_month) ||
                (month == current_month && day > current_day) ||
                (month == current_month && day == current_day && hour > current_hour)) {
                // future month/day/hour -> use previous year
                year--;
            }
            provided_tm.tm_year = year;
            provided_time = std::mktime(&provided_tm);
        } else {
            log_msg("The provided time is within the last 24 hours. Using system time.");
            sys_time = true;
            provided_time = now;
        }
    }

    // now convert provided time to UTC
    auto ptime_utc_tm = *std::gmtime(&provided_time);
    char buffer[500];
    strftime(buffer, sizeof(buffer), "%Y-%m-%d-%H:%M:%S", &ptime_utc_tm);
    log_msg("provided time (UTC): %s", buffer);

    auto [url, ctime_utc_tm, cycle] = GetDownloadUrl(sys_time, ptime_utc_tm);

    // Get grib file's date date in yyyy-mm-dd format
    strftime(buffer, sizeof(buffer), "%Y-%m-%d", &ctime_utc_tm);
    char fn_buffer[500];
    snprintf(fn_buffer, sizeof(fn_buffer), "%s/%s_%d_noaa.grib2", output_dir.c_str(), buffer, cycle);
    std::string grib_file_path = fn_buffer;
    log_msg("GRIB file path: '%s'", grib_file_path.c_str());

    // if file does not exist, download
    if (!std::filesystem::exists(grib_file_path)) {
        log_msg("Downloading GRIB file from '%s'", url.c_str());

        FILE *out = fopen(grib_file_path.c_str(), "wb");
        if (out == NULL) {
            log_msg("ERROR: can't create '%s'", grib_file_path.c_str());
            return "";
        }
        if (HttpGet(url.c_str(), out, 10))
            log_msg("GRIB File downloaded successfully");
        else {
            log_msg("GRIB File download failed");
            grib_file_path = "";
        }

        fclose(out);
    }

    return grib_file_path;
}

static void
RemoveOldGribFiles(std::string file_to_keep)
{
    // posixify filenames for textual comparison
#if IBM == 1
    std::replace(file_to_keep.begin(), file_to_keep.end(), '\\', '/');
#endif

    log_msg("Removing old grib files");
    log_msg("File to keep: %s", file_to_keep.c_str());
    log_msg("Grib file folder: %s", output_dir.c_str());

    try {
        for (const auto& entry : std::filesystem::directory_iterator(output_dir)) {
            auto path = entry.path().string();
#if IBM == 1
            std::replace(path.begin(), path.end(), '\\', '/');
#endif
            // Check for files with .grib extension
            if (path.find("_noaa.grib2") != std::string::npos && path.find(file_to_keep) == std::string::npos) {
                std::filesystem::remove(path);
                log_msg("Removed: %s", path.c_str());
            }
        }
    } catch (const std::exception& e) {
        log_msg("Error removing old grib files: %s", e.what());
    }
}

// everything is synchronously fired by the flightloop so we don't need mutexes
static bool download_active;
static std::future<bool> download_future;

static const char *wgrib2 =
#if IBM == 1
"/WIN32wgrib2.exe";
#elif LIN == 1
"/linux-wgrib2";
#elif APL == 1
"/OSX11wgrib2";
#endif

// Runs async
static bool
DownloadAndProcessGribFile(bool sys_time, int day, int month, int hour)
{
    const char *snod_csv_name = std::getenv("USE_SNOD_CSV");

    if (NULL == snod_csv_name) {
        std::string grib_file_path = DownloadGribFile(sys_time, day, month, hour);
        if (grib_file_path.size() == 0)
            return false;

        snod_csv_name = "snod.csv";

        // export grib file to csv
        // 0:3600:0.1 means scan longitude from 0, 3600 steps with step 0.1 degree
        // -90:1800:0.1 means scan latitude from -90, 1800 steps with step 0.1 degree
        std::string cmd = plugin_dir + "/bin" + wgrib2
            + " -s -lola 0:3600:0.1 -90:1800:0.1 " + snod_csv_name + " spread " + grib_file_path + " -match_fs SNOD";

        log_msg("cmd:'%s'", cmd.c_str());
        int ex = sub_exec(cmd);
        if (ex != 0)
            return false;

        RemoveOldGribFiles(grib_file_path);
    } else
        log_msg("Using existing snod_csv file '%s'", snod_csv_name);

    grib_snod_map->load_csv(snod_csv_name);
    ElsaOnTheCoast(*grib_snod_map, *snod_map);
    CreateSnowMapPng("snow_depth.png");
    return true;
}

static bool
DownloadAndProcess(bool sys_time, int day, int month, int hour)
{
    for (int i = 0; i < 3; i++) {
        bool res = DownloadAndProcessGribFile(sys_time, month, day, hour);
        if (!res) {
            log_msg("Download grib file failed, retry: %d", i);
        } else {
            log_msg("Download and process grib file successfully");
            return true;
        }
    }

    log_msg("grib download/process: all retry failed");
    return false;
}

// start download in the background
void
StartAsyncDownload(bool sys_time, int day, int month, int hour)
{
    if (download_active) {
        log_msg("Download is already in progress, request ignored");
    }

    download_future = std::async(std::launch::async, DownloadAndProcess, sys_time, day, month, hour);
    download_active = true;
}

// return true if do
bool
CheckAsyncDownload()
{
    if (!download_active)
        return false;

    if (std::future_status::ready != download_future.wait_for(std::chrono::seconds::zero()))
        return false;

    download_active = false;
    bool res = download_future.get();
    log_msg("Download status: %d", res);
    return true;
}
